File: read-csv-php/sales.php
<?php
$start_time = microtime(true);

$file = fopen("../generate-csv/sales_data.csv", "r");
$total_sales = 0;
$product_sales = [];

fgetcsv($file); // Skip header
while (($line = fgetcsv($file)) !== false) {
    $product_id = $line[1];
    $quantity = (int)$line[2];
    $price = (float)$line[3];
    $total = $quantity * $price;

    $total_sales += $total;
    if (!isset($product_sales[$product_id])) {
        $product_sales[$product_id] = 0;
    }
    $product_sales[$product_id] += $total;
}
fclose($file);

arsort($product_sales);
$top_product = array_key_first($product_sales);

$end_time = microtime(true);
$execution_time = ($end_time - $start_time);

echo "PHP Execution time: ".$execution_time." seconds\n";
echo "Total Sales: $".$total_sales."\n";
echo "Top Product: ".$top_product." with sales $".$product_sales[$top_product]."\n";

File: read-csv-go/sales.go
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"strconv"
	"time"
)

func main() {
	start := time.Now()

	file, err := os.Open("../generate-csv/sales_data.csv")
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	defer file.Close()

	reader := csv.NewReader(file)
	_, _ = reader.Read() // Skip header

	totalSales := 0.0
	productSales := make(map[string]float64)

	var topProduct string
	var topSales float64

	for {
		line, err := reader.Read()
		if err != nil {
			break
		}

		productID := line[1]
		quantity, _ := strconv.Atoi(line[2])
		price, _ := strconv.ParseFloat(line[3], 64)
		total := float64(quantity) * price

		totalSales += total
		productSales[productID] += total

		// Update topProduct and topSales in the same loop
		if productSales[productID] > topSales {
			topProduct = productID
			topSales = productSales[productID]
		}
	}

	elapsed := time.Since(start)
	fmt.Printf("Golang Execution time: %s\n", elapsed)
	fmt.Printf("Total Sales: $%.2f\n", totalSales)
	fmt.Printf("Top Product: %s with sales $%.2f\n", topProduct, topSales)
}
